Fast path for binary ops that bypass tensor iterator.

kimishpatel opened the pull request:
Summary:
Changes add fast path to bypass tensor iterator when tensors:
- are on cpu.
- are contig.
- have the same number of dimensions.
- each dimension has the same size.
- are of the same type.

Results on top of master using framework overhead benchmar:
Experimented with 3 sizes: 1, 16, 64. All 1 dim tensors.
In case of broadcast, the second tensor of the add was of 1 element.

On master:
add_tensors_loop:Num Operands=2_[1],Graph mode:True, latency per iter (us):1.7277789115905762
add_tensors_loop:Num Operands=2_[16],Graph mode:True, latency per iter (us):1.8338178396224976
add_tensors_loop:Num Operands=2_[64],Graph mode:True, latency per iter (us):1.8206777572631836
On master with broadcast:
add_tensors_loop:Num Operands=2_[1],Graph mode:True, latency per iter (us):1.711668610572815
add_tensors_loop:Num Operands=2_[16],Graph mode:True, latency per iter (us):1.8416098356246948
add_tensors_loop:Num Operands=2_[64],Graph mode:True, latency per iter (us):1.895913004875183

===========

With fast path:
add_tensors_loop:Num Operands=2_[1],Graph mode:True, latency per iter (us):1.2124907970428467
add_tensors_loop:Num Operands=2_[16],Graph mode:True, latency per iter (us):1.2013496160507202
add_tensors_loop:Num Operands=2_[64],Graph mode:True, latency per iter (us):1.2547712326049805
Fast path with broadcast:
add_tensors_loop:Num Operands=2_[1],Graph mode:True, latency per iter (us):1.2107170820236206
add_tensors_loop:Num Operands=2_[16],Graph mode:True, latency per iter (us):2.0033864974975586
add_tensors_loop:Num Operands=2_[64],Graph mode:True, latency per iter (us):1.9163089990615845

Non broadcast case fast path saves about 500 to 600ns. Adding fast path has some overhead when we don't end up taking the fast path as in the case of broadcast. This is about 50-150ns.

Differential Revision: D16359346

VitalyFedyunin commented:
If you want to continue exploring this way:

`/var/lib/jenkins/workspace/build/aten/src/ATen/native/cpu/FastPathBinaryOpsKernelCPU.cpp.DEFAULT.cpp:37: multiple definition of `at::native::all_cpu_fast_path_conds_satisfied(at::Tensor const&, at::Tensor const&)'` is the legit error

But as alternative I recommend to explore the option to modify TensorIterator compute_shape and capture if all shapes are the same, having this information you can shortcut most of the loops and end-up with equal performance gain but cleaner operators code.

kimishpatel commented:
@VitalyFedyunin, let me look through it and see, but my gut feeling is that there is a lot of wrapping and unwrapping that happens with tensoriterator, perhaps via serial_for_each. Let me try out what you suggested. 
Thanks for the pointer to all_cpu_fast_path_conds_satisfied. I need to move that out of cpu dir.

kimishpatel commented:
@VitalyFedyunin, I looked into if we something can be saved knowing shape etc, but it is not straighforward. Sam already did something for reorder_dimensions. Tensor iterator needs some of the information in order to iterate through and provide appropriate data pointers to the lambdas. We can try to add a fast path inside tensor iterator such that it takes an entirely different path (such as avoiding builder step, and not iterating via serial_for_each and packing data pointers), but that won't be very different from what is there. What do you think?

kimishpatel committed:
Temp comming

fbshipit-source-id: 7bf0e6c6cfe70d03b1c5b82086c625515fbcdb05

kimishpatel committed:
Added a way to have one tensor whose shape has at least one dim with size 1

fbshipit-source-id: 799df5dc2be12e0f7d729f5b1ba159a09bcd53e7

kimishpatel committed:
Fast path for binary ops that bypass tensor iterator. (#23167)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/23167

Changes add fast path to bypass tensor iterator when tensors:
- are on cpu.
- are contig.
- have the same number of dimensions.
- each dimension has the same size.
- are of the same type.

Results on top of master using framework overhead benchmar:
Experimented with 3 sizes: 1, 16, 64. All 1 dim tensors.
In case of broadcast, the second tensor of the add was of 1 element.

On master:
add_tensors_loop:Num Operands=2_[1],Graph mode:True, latency per iter (us):1.7277789115905762
add_tensors_loop:Num Operands=2_[16],Graph mode:True, latency per iter (us):1.8338178396224976
add_tensors_loop:Num Operands=2_[64],Graph mode:True, latency per iter (us):1.8206777572631836
On master with broadcast:
add_tensors_loop:Num Operands=2_[1],Graph mode:True, latency per iter (us):1.711668610572815
add_tensors_loop:Num Operands=2_[16],Graph mode:True, latency per iter (us):1.8416098356246948
add_tensors_loop:Num Operands=2_[64],Graph mode:True, latency per iter (us):1.895913004875183

===========

With fast path:
add_tensors_loop:Num Operands=2_[1],Graph mode:True, latency per iter (us):1.2124907970428467
add_tensors_loop:Num Operands=2_[16],Graph mode:True, latency per iter (us):1.2013496160507202
add_tensors_loop:Num Operands=2_[64],Graph mode:True, latency per iter (us):1.2547712326049805
Fast path with broadcast:
add_tensors_loop:Num Operands=2_[1],Graph mode:True, latency per iter (us):1.2107170820236206
add_tensors_loop:Num Operands=2_[16],Graph mode:True, latency per iter (us):2.0033864974975586
add_tensors_loop:Num Operands=2_[64],Graph mode:True, latency per iter (us):1.9163089990615845

Non broadcast case fast path saves about 500 to 600ns. Adding fast path has some overhead when we don't end up taking the fast path as in the case of broadcast. This is about 50-150ns.

Differential Revision: D16359346

fbshipit-source-id: 4c5db703c4b0ed2a8ced652fc3ca77dbbb919d82

dzhulgakov commented:
If we add it as a fast pass in `TensorIterator` + `cpu_kernel_vec`, it'd benefit all operations implemented through them. In the ideal case, we'd just substitute the same lambda passed `cpu_kernel_vec` inside the loop and obtain the fast contiguous kernel without having to change per-operator code.

If I recall correctly, from @zheng-xq's tests modern compilers are smart enough to inline the lambda. But we should check.

VitalyFedyunin commented:
> If we add it as a fast pass in `TensorIterator` + `cpu_kernel_vec`, it'd benefit all operations implemented through them. In the ideal case, we'd just substitute the same lambda passed `cpu_kernel_vec` inside the loop and obtain the fast contiguous kernel without having to change per-operator code.

It is already happening inside of the TensorIterator. Observed slow down related to dimensions reordering, collapsing etc, more universal shape checking and things like that. I'm currently looking if we can elegantly bypass some smallvec operations.

apaszke commented on aten/src/ATen/native/FastPathConditions.h:
```diff
@@ -0,0 +1,34 @@
+      return false;
+    }
+  }
+  return true;
```
This whole function is equivalent to `t1.sizes() == t2.sizes()`

apaszke commented on aten/src/ATen/native/FastPathConditions.h:
```diff
@@ -0,0 +1,34 @@
+bool enable_contiguous_eq_size_cpu_fastpath(const Tensor& t1, const Tensor& t2) {
+  return all_dims_are_same(t1, t2) && t2.is_contiguous() &&
+         t1.device().is_cpu() && t2.device().is_cpu() &&
+         t1.is_contiguous() &&
```
I'd move the `is_cpu` checks to the front, because they're likely to be cheap and can let us quickly move on to the generic path on other devices.

kimishpatel replied:
@apaszke, I actually had it that way before, but I was trying to reduce the overhead of fastpath condition checking. Since we want fastpath for inference where it is likely that the tensors are on cpu most likely those conditions will hold true. Thus I moved the condition which are less likely to be true to the front. I havent gotten around to digging into what is contributing to the overhead of condition checking, when we dont go through the fast path. So this is one place I will explore to see how to get out of here quickly.

apaszke replied:
This is not what you should optimize for. If your assumption is that you can use the fast path, then the order of those checks is actually irrelevant, because they will all have to be executed anyway. What's important is which conditions are the ones which will fail the fastest if you _can't_ use the fast path (because then you'll get the short-circuit very quickly). Hence, it's better to put the cheap and quickly failing ones first.

I understand the desire to make inference fast, but you have to remember that this library has a thousand other users with other constraints.

VitalyFedyunin commented:
@ngimel landed TensorIterator friendly implementation of this idea.
